src/ingestion/kafka-queue.ts

```diff
@@ -39,27 +37,29 @@ export class KafkaQueue implements Queue {
         isStale,
     }: EachBatchPayload): Promise<void> {
         const batchProcessingTimer = new Date()
-        const rawEvents: RawEventMessage[] = batch.messages.map((message) => ({
-            ...JSON.parse(message.value!.toString()),
-            kafka_offset: message.offset,
-        }))
-        const parsedEvents = rawEvents.map((rawEvent) => ({
-            ...rawEvent,
-            data: JSON.parse(rawEvent.data),
-        }))
-        const pluginEvents: PluginEvent[] = rawEvents.map((rawEvent) => {
-            const { data: dataStr, ...restOfRawEvent } = rawEvent
-            const event = { ...restOfRawEvent, ...JSON.parse(dataStr) }
+
+        const uuidOrder = new Map<string, number>()
+        const uuidOffset = new Map<string, string>()
+        const pluginEvents = batch.messages.map((message, index) => {
+            const { data: dataStr, ...rawEvent } = JSON.parse(message.value!.toString())
+            const event = { ...rawEvent, ...JSON.parse(dataStr) }
+            uuidOrder.set(event.uuid, index)
+            uuidOffset.set(event.uuid, message.offset)
             return {
                 ...event,
-                kafka_offset: restOfRawEvent.kafka_offset,
                 site_url: event.site_url || null,
                 ip: event.ip || null,
             }
         })
-        const processedEvents: PluginEvent[] = (
-            await this.processEventBatch(pluginEvents)
-        ).filter((event: PluginEvent[] | false | null | undefined) => Boolean(event))
+
+        const processedEvents = await this.processEventBatch(pluginEvents)
+
+        // Sort in the original order that the events came in, putting any randomly added events to the end.
+        // This is so we would resolve the correct kafka offsets in order.
+        processedEvents.sort(
+            (a, b) => (uuidOrder.get(a.uuid!) || pluginEvents.length) - (uuidOrder.get(b.uuid!) || pluginEvents.length)
+        )
```

Comment:
Nice that the tacked on events are last!
Though as for the non-null assertion, as far as I understand, `processEventBatch` can return an array of practically anything, including non-objects falsy or truthy. 🤔

Reply:
It should always return this or crash: https://github.com/PostHog/plugin-server/pull/124/files#diff-5d37292cf21755714262793e7553ee317a12b157d0601e27a0ca26e096a60c00L267

Though, yeah, there's no real sanitization. The plugins could return just `{ "n00b": "lololololol" }` as an event and it would make its way here...

Reply:
I suppose the `get` is going to return `undefined` in any such case, so this should be fine even if the assertion is too optimistic.
Side note, I think we make this assumption in some more dangerous places too, so probably will have to take a look at some event structure validation in another PR.